asf_heat_pump_affordability/notebooks/identify_and_review_analytical_sample.py

```diff
@@ -0,0 +1,338 @@
+# We will see if the distribution of costs in removed rows is the same as that of rows which have been kept in the sample for analysis.
+
+# %%
+dropped_rows = filter_by_year.loc[~filter_by_year.index.isin(sample.index)]
+assert (len(filter_by_year) - len(sample)) == len(dropped_rows)
+
+# %%
+# In this case we'll normalise the data so they're comparable.
+
+fig, ax = plt.subplots(nrows=1, ncols=2, sharey=True, figsize=(10, 3))
+ax[0].hist(sample["cost"], density=True)
+ax[0].set_ylabel("Count")
+ax[1].hist(dropped_rows["cost"], density=True)
+for axis, title in zip(ax, ["kept", "dropped"]):
+    axis.set_title(f"Distribution of installation cost in {title} rows")
+    axis.set(xlabel="Installation cost, £")
+plt.show()
+
+# %%
+# There's not much benefit in doing this, suggest dropping.
+
+fig, ax = plt.subplots(nrows=1, ncols=2, sharey=True, figsize=(10, 3))
+ax[0].hist(sample["cost"], log=True)
+ax[0].set_ylabel("Count (log scale)")
+ax[1].hist(dropped_rows["cost"], log=True)
+for axis, title in zip(ax, ["kept", "dropped"]):
+    axis.set_title(f"Distribution of installation cost in {title} rows")
+    axis.set(xlabel="Installation cost, £")
+plt.show()
+
+# %% [markdown]
+# ## Note on comparison
+#
+# In this particular case, I'm interested in how similar/different the dropped and retained groups are in terms of cost.
+#
+# I also am not interested in the water source/ground source heat pumps at all, intuitively I know that they'll inflate the mean cost and give me a distribution with a heavier tail. Really, my key concern is: do costs for ASHP look different for installs where we don't have an EPC?
+#
+# If costs are similar, I can be relatively happy that the sample isn't massively missing something.
+#
+# I can do this in a couple of ways - 1. with statistical summaries, and 2. with visualisation.
+#
+# Descriptively the dropped rows and sample look similar - similar means, dropped rows has a higher std. deviation, which is interesting, sample again a bit higher at each quartile.
+#
+# Let's see what this actually looks like graphed out. NB I'm going to truncate the cost variable at £30,000 which is roughly the 99th percentile of the distribution, as I'm not actually that interested in the tail.
+#
+# Ultimately, I can see there is a slight trend for sampled costs to be a bit higher. However, I can't know whether this is a 'true' difference, or a difference in composition of the sample. Ultimately, I'm relatively happy that the sample isn't that different from the dropped rows in cost terms.
+
+# %%
+# filter for just ASHP in dropped rows.
+dropped_rows = dropped_rows.loc[lambda df: df["tech_type"] == "Air Source Heat Pump", :]
+
+# %%
+dropped_rows["cost"].describe()
+
+# %%
+sample["cost"].describe()
+
+# %%
+f, ax = plt.subplots(figsize=(8, 5))
+
+ax.hist(
+    dropped_rows.loc[lambda df: df["cost"] < 30_000, "cost"],
+    density=True,
+    histtype="bar",
+    alpha=0.5,
+    bins=20,
+    color="indianred",
+    label="Dropped",
```

Comment:
``
```suggestion
binned_data, bin_defs = pd.cut(sample[sample['cost'] < 30000]['cost'], bins=20, retbins=True)
f, ax = plt.subplots(figsize=(8, 5))

ax.hist(
    dropped_rows.loc[lambda df: df["cost"] < 30_000, "cost"],
    density=True,
    histtype="bar",
    alpha=0.5,
    bins=bin_defs,
    color="indianred",
    label="Dropped",
```
Same comment for my histogram above. Is it better to use the same bin ranges for each dataset for better comparison? It looks like it has negligible effect here because the data is truncated but is noticeable on non-truncated data.

Reply:
Ideally you should apply the same bin ranges. I'm comfortable here that it doesn't really matter.